### ISSUE TITLE
Make Bluetooth RPC work in the Mojo/Android world

### ISSUE THREAD

**Issue by aghassemi** [vanadium/issues#1061]
Use-case is Mojo Syncbase being able to sync with other devices over Bluetooth. Ideally the solution extends transparently to any other service and not just Syncbase.

Maybe we can resurrect https://cs.corp.google.com/#piper///depot/veyron/v/src/veyron/runtimes/google/lib/bluetooth/bt.c&type=cs and ensure it works in Android and Linux so we don't need Java

Maybe Mojo should provide a Bluetooth service like they do with Network services.

Maybe we can wrap the Java Bluetooth RPC code in a mojo service and call it from Go.

@spetrovic77 @bjornick @asadovsky @kenashcraft @mattr-google

**Comment by ghost** [comment]
So, we can easily make it work on desktop linux machines, but it's not clear that's of use to anyone.  We have an implementation that works if you are using java/android, but it can't be used by pure go programs on android because android doesn't expose the requisite apis.

We would have to have a mojo service to do bluetooth.  Do you know if that is being worked on?  Does it have a bug?  After that I (or you if you want) can make a vanadium protocol that wraps it.

For now I'll assign this to myself (since all bugs should be assigned) but if you are going to work on it, please feel free to reassign.

**Comment by joeblew99** [comment]
its of use to me :)

I am using Vanadium as a prototype. So far its pretty nice to be able to get so much done with just golang. This is an awesome project.

I need BT access on IOS and Android
I also need the same on Linux (currently running in NUC's and RaspberryPis.

I am currently using c codebase for reading RFID tags for now, as i need that too. However, i am still working through the hardware tag side of things; for example there are RFID tags that can be read by BT devices for example. Even printer paper these days with them embedded. The reason this is awesome is because RFID tags are 10 to 30 times cheaper, you can scan 100's of rfid tags with less than 1 % error rate, but you can still read them with bluetooth as a consumer from your mobile

I just wanted to let you know it IS useful.

cheers

Joe

**Comment by ghost** [comment]
Hello,

This bug is particularly for Mojo, are you planning to use Mojo?

One problem is that as far as we can see there is no general access to Bluetooth on IOS devices.

See for example:
http://stackoverflow.com/questions/18884705/transfer-data-between-ios-and-android-via-bluetooth
http://www.themethodology.net/2015/03/how-to-use-classic-bluetooth-on-ios.html

We could implement RPC over BLE, but then the performance will be very poor and there are some severe restrictions, for example the "advertiser" in BLE (which would most naturally be the server) can only have one connection at a time.

We already have Bluetooth on Android, but it requires that you write part of your app in Java since the Bluetooth stack must be initialized using the Java APIs.  We can add desktop linux (ie. for raspberry pi)  Feel free to add a new issue for that.

I don't know if Mojo will implement a Bluetooth classic service given that IOS is out of reach, and if they provide a BLE service it's not clear it will be targeted at generic byte transport since that doesn't seem to be a common use case.  For now I think this bug will remain blocked on Mojo.

**Comment by joeblew99** [comment]
@mattr-google 
I can plan to use Mojo is needed. If i have to then i will if that gets the job done.

I read the link. Lots of people hitting this nice Apple road block..
I heard that Apple had made BT access restricted, but guessed it would be for maybe BT indoor positioning, not standard BT access. Huge pain for trying to just get things done eh...

One idea ? Maybe sucks...
It sounds like making 2 devices talk to each other then might have to be faked out of bounds. In other works they both talk over wifi via a third party. This is just a work around !
With Vanadium the two devices can find each other even when each is heavily NATed ?

RPC over BT....
Perf- not a biggy for me, because we are only transferring strings of maybe 100 chars, like ID numbers etc.
Single Connection session - also fine for me, because each device will only ever be talking to another single device per session.

BT Android..
Can live with coding that part in java for now. I wonder what the Flutter guys have in store here ? From my world i can talk to the BT stakc from the Dart / Flutter world OR the GO / Vanadium world.  I have no idea though where flutter / Fletch is on this. I wil ask them after this... 

Thanks in advance for the all the info. I can see that BT / BTE is damn hard to get coverage across different devices... I guess this is why i raised the aspect of out of bounds communication with wifi, and a third party server in the middle. Would be interesting to hear your thouhts on this in realtion to vanadium.
In the use cases i am looking at, i always have a Server that is inside the same NATed network where the other devices running Vanadium are located. SO for me its certainly do.

Also if anyone has any idea on the state of play NFC with Vanadium / GO on android and IOS it would really help me.

**Comment by ghost** [comment]
Having devices talk through a proxy is already supported, and is a good way around NAT issues.  We also have a discovery service which can help devices find each other.  I would say if both devices have access to WiFi we have pretty good support.

There has been some experimental work using NFC to transfer credentials, but I don't think there's been anything more general.

**Comment by joeblew99** [comment]
Thanks. I am using golang for talking to RFID and NFC on embedded hardware. It's very easy so far.
But that's because I got to choose the hardware.

I guess its a matter of trying it. Like everything I guess you guys have to prioritise what sensor APIs you wrap with golang from the ndk layer ?

**Comment by ghost** [comment]
@asimshankar Has more info about the ongoing NFC work.

**Comment by asimshankar** [comment]
I'm curious about what hardware and Golang libraries you're using for RFID and NFC on the Pis.

We don't have many plans when it comes to NFC integration with Android, but you might find [this](https://docs.google.com/document/d/12jVwksEdyWbDVSXB6ffR24dIl18moGovxZKKN5zrIlA/edit?usp=sharing) interesting

**Comment by joeblew99** [comment]
No libs. Just raspberry and talking to their hardware direct. I might be
able to share the code too. Have to check with management

https://www.adafruit.com/products/364

Thanks for the doc.

**Comment by ghost** [comment]
I'm going to close this for now.  We don't have immediate plans to do more RPC integration work with Mojo.
